Add possibility to fake function with temp or derived tables

dkultasev opened the pull request:
This pull requests adds possibility to fake inline or multi-statement functions without creating additional functions. It supports 2 new ways of faking it:
* Use pre-populated table
* Use derived table

Examples:

```sql
CREATE FUNCTION dbo.VeryImportantFunction (@param INT)
RETURNS TABLE AS RETURN (SELECT 1 AS one);
GO

BEGIN TRAN
 CREATE TABLE #expected (one INT);
 INSERT INTO #expected VALUES (1), (2);

 EXEC tSQLt.FakeFunction @FunctionName = N'dbo.VeryImportantFunction', 
                         @FakeDataSource = N'(VALUES (1), (2)) a(one)'
 
 SELECT * INTO #actual FROM dbo.VeryImportantFunction(0);

 EXEC tSQLt.AssertEqualsTable @Expected = N'#expected', 
     @Actual = N'#actual' 
 
ROLLBACK

BEGIN TRAN
 CREATE TABLE #expected (one INT);
 INSERT INTO #expected VALUES (1), (2);

 EXEC tSQLt.FakeFunction @FunctionName = N'dbo.VeryImportantFunction', 
                         @FakeDataSource = N'SELECT 1 AS one UNION ALL SELECT 2'
 
 SELECT * INTO #actual FROM dbo.VeryImportantFunction(0);

 EXEC tSQLt.AssertEqualsTable @Expected = N'#expected', 
     @Actual = N'#actual' 
 
ROLLBACK


BEGIN TRAN
 CREATE TABLE #expected (one INT);
 INSERT INTO #expected VALUES (1), (2);

 EXEC tSQLt.FakeFunction @FunctionName = N'dbo.VeryImportantFunction', 
                         @FakeDataSource = N'#expected'
 
 SELECT * INTO #actual FROM dbo.VeryImportantFunction(0);

 EXEC tSQLt.AssertEqualsTable @Expected = N'#expected', 
     @Actual = N'#actual' 
 
ROLLBACK
```

lizbaron reviewed (changes requested):
Hey Dmitrij,

This looks like great functionality! Sebastian and I reviewed it and validated the PR on 2019 and 2008R2. However, we do have some feedback we'd like you to address before we test it on all supported versions. 

I've listed the requested modifications by file.

- Source/tSQLt.FakeFunction.ssp.sql
  - Newline required on last line
- Source/tSQLt.Private_CreateFakeFunction.ssp.sql
  - Newline required on last line
- Source/tSQLt.Private_PrepareFakeFunctionOutputTable.ssp.sql
  - Can you say more about why we are creating a completely new output table? What is the issue with using the original @FakeDataSource? It seems like the logic around the @FakeDataSource variable could be moved into the CreateFakeFunction without creating a new table.
  - If we really do need the output table, it should have a schema specified.
  - Newline required on last line
- Source/tSQLt.Private_ValidateObjectsCompatibleWithFakeFunction.ssp.sql
  - Line 28: Please modify the error message to `RAISERROR ('Both @FakeFunctionName and @FakeDataSource are valued. Please use only one.', 16, 10);`
  - Line 37: Please modify the error message to `RAISERROR('You can use @FakeDataSource only with Inline or Multi-Statement Table-Valued functions.', 16, 10);`
  - Newline required on last line
- Tests/FakeFunctionTests.class.sql
  - In general test names should reflect desired outcome, not the situation tested. Please see the requested renames for examples. 
  - The following test seems to not add value specific to the @FakeDataSource functionality as the functionality tested looks like it is already tested in tSQLt.Private_GetFullTypeName
    - `CREATE PROCEDURE FakeFunctionTests.[test @FakeDataSource fakes function with all possible standard types]`
  - Please update tests
    - `CREATE PROCEDURE FakeFunctionTests.[test @FakeDataSource to accept function with single param]`
      - Rename to “test can fake function with one parameter”
      - Delete line 486 CREATE TABLE #actual (a INT)
      - Update line 490 to SELECT * INTO syntax
    - `CREATE PROCEDURE FakeFunctionTests.[test @FakeDataSource to accept function with multiple params]`
      - Rename to “test can fake function with multiple parameters”
      - Update to include three parameters
      - Delete line 503
      - Update line 507 to SELECT * INTO syntax
    - `CREATE PROCEDURE FakeFunctionTests.[test @FakeDataSource with temp table with inline table function]`
      - Rename to “test can fake Inline table function using a temp table as fake data source”
    - `CREATE PROCEDURE FakeFunctionTests.[test @FakeDataSource fakes function with VALUES clause]`
      - This is a good name but would prefer a rename to “test can fake function with VALUES clause as fake data source”
  - Need test to check CLR table-valued functions with @FakeDataSource
  - Need to test not just temp tables, but also tables with two-part names.
  - Newline required on last line

dkultasev commented:
@lizbaron how to test the changes now? I can't run the build locally after merge with master

lizbaron reviewed (changes requested):
Hey, Dmitrij sorry about the confusion around running the build. We changed from using NAnt to Ant. You can see all the parameters required if you look at tsqlt/Build/LocalBuild.bat

Thanks for the explanation on the new tables. It sounds like we really need to keep `Source/tSQLt.Private_PrepareFakeFunctionOutputTable.ssp.sql` In that case I have additional comments.
  - I don’t like tSQLt creating dbo objects, so please either use the tSQLt schema or follow a similar naming convention to the SpyProcedure function. Sort of like: `functionschema.functionname_tsqltoutputtable` Both strategies have the advantage that it will be clear that these objects can be safely deleted if encountered outside of a test.
  - I would like you to remove `tSQLt.SupressOutput` from line 22, or if you think it is required, please write a test for it.
  - And finally, please create tests for the logic in the “if” statement on lines 12-14. At a minimum there needs to be a test that fails if you delete line 13.

For the CLR tests, please find examples in `Tests/FakeFunctionTests.class.sql` using `tSQLt_testutil.AClrTvf`and `tSQLt_testutil.AClrSvf`. Please make sure that you test iterations of both. We would expect Svf to throw similar errors as a normal scalar valued function does, just in case that isn’t clear.

This is great work, and we look forward to merging it soon.

dkultasev commented:
I pushed the changes for all the points except the tests with CLR done. For CLR I'll need to setup the Ant first to be able to build the project. That would take some time, most probably I'll do that on weekend.

mbt1 commented:
Great. Looking forward to seeing the rest of your changes. BTW, just remembered, there is a function in tSQLt the generates names for temporary objects and deals with collisions:

`SELECT @ResultTable = tSQLt.Private::CreateUniqueObjectName();`

We should probably use that for your table. (Sorry about the back and forth here.)

dkultasev commented:
I've committed the change with `tSQLt.Private::CreateUniqueObjectName();`. Could you please help me with the local build setup?
I've changed call arguments in the `External Tools`:
* Changed `Command` to `$(SolutionDir)\Build\LocalBuild.bat`
* Changed `Arguments` to `%AntHome% %Net4Home% anyValue Dev_tSQLt tSQLt_dev`
* Changed `Initial directory` to `$(SolutionDir)`

When I run the build from _SSMS_ I'm getting: 
```
+-------------------------+
: Executing Local Build   :
+-------------------------+
Parameters:
AntHome: "C:\Users\dkult\OneDrive\Apps\apache-ant-1.9.15"
NET4Home: "C:\Windows\Microsoft.NET\Framework64\v4.0.30319\"
SQLCMDPath: "anyValue"
SQLVersion: "deprecated"
SQLInstanceName: "Dev_tSQLt"
DBName: "tSQLt_dev"
+-------------------------+
: Starting CLR BUILD      :
+-------------------------+
Buildfile: C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\tSQLt.buildCLR.xml

validate.parameters:

validate.parameter:
     [echo] C:\Windows\Microsoft.NET\Framework64\v4.0.30319"

init:
   [delete] Deleting directory C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\temp
    [mkdir] Created dir: C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\temp
    [mkdir] Created dir: C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\temp\tSQLtCLR

clr.compile:
     [echo] Compile CLR Command: "C:\Windows\Microsoft.NET\Framework64\v4.0.30319"\msbuild.exe" tSQLtCLR.csproj /p:Configuration=CruiseControl /nologo
     [exec] The system cannot find the path specified.

BUILD FAILED
C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\tSQLt.buildCLR.xml:50: exec returned: 1

Total time: 0 seconds
+-------------------------+
:  !!! BUILD FAILED !!!   :
+-------------------------+

```

dkultasev commented:
I've committed the change with `tSQLt.Private::CreateUniqueObjectName();`. Could you please help me with the local build setup?
I've changed call arguments in the `External Tools`:
* Changed `Command` to `$(SolutionDir)\Build\LocalBuild.bat`
* Changed `Arguments` to `%AntHome% %Net4Home% anyValue Dev_tSQLt tSQLt_dev`
* Changed `Initial directory` to `$(SolutionDir)`

When I run the build from _SSMS_ I'm getting: 
```
+-------------------------+
: Executing Local Build   :
+-------------------------+
Parameters:
AntHome: "C:\Users\dkult\OneDrive\Apps\apache-ant-1.9.15"
NET4Home: "C:\Windows\Microsoft.NET\Framework64\v4.0.30319\"
SQLCMDPath: "C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\sqlcmd.exe"
SQLVersion: "deprecated"
SQLInstanceName: "Dev_tSQLt"
DBName: "tSQLt_dev"
+-------------------------+
: Starting CLR BUILD      :
+-------------------------+
Buildfile: C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\tSQLt.buildCLR.xml

validate.parameters:

validate.parameter:
     [echo] C:\Windows\Microsoft.NET\Framework64\v4.0.30319"

init:
   [delete] Deleting directory C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\temp
    [mkdir] Created dir: C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\temp
    [mkdir] Created dir: C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\temp\tSQLtCLR

clr.compile:
     [echo] Compile CLR Command: "C:\Windows\Microsoft.NET\Framework64\v4.0.30319"\msbuild.exe" tSQLtCLR.csproj /p:Configuration=CruiseControl /nologo
     [exec] The system cannot find the path specified.

BUILD FAILED
C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\Build\tSQLt.buildCLR.xml:50: exec returned: 1

Total time: 0 seconds
+-------------------------+
:  !!! BUILD FAILED !!!   :
+-------------------------+

```

mbt1 commented:
If I read the error right, it can't find msbuild.exe

Could you confirm that your %Net4Home% is set correctly? For me the three path parameters are:

```
>ECHO "%AntHome%" "%NET4Home%" "%SQLCMDPath%"
"C:\tools\apache-ant-1.10.6" "C:\Windows\Microsoft.NET\Framework\v4.0.30319" "C:\Program Files\Microsoft SQL Server\110\Tools\Binn"
```

In particular, I don't have the 64 after the word Framework.

----

After you got that working, you'll run into a certificate issue. To fix that, you'll need to use visual studio to create a signing key (project properties->signing->new key) and then use one of the two batch files in the OfficialSigningKey directory to install the generated key on your machine. (You don't need to save the project that you used to create the signing key, just the key itself.)

dkultasev commented:
My possible problem is that the Visual Studio solution is not built. Could you please provide me instructions how to build the solution? Every time I need to setup the build locally I have so much troubles dealing with the certificates ...
By the way, my Visual Studio version is 2019. Is that OK?

lizbaron commented:
Hi @dkultasev,

Sebastian is temporarily unavailable, but did you try what he suggested? The tSQLt build takes care of building the Visual Studio solution. There is no additional step required. And it looks like the error is pretty clear, and is specific to your local setup. 

If that is not where you are stuck anymore, can you please give us details on:
- What are you trying to do?
- What problem did you encounter?
- How have you tried to fix it?

Let me know!
Liz

dkultasev commented:
Hi Liz, sorry I was very busy this week. Will work on that on weekend

пт, 30 окт. 2020 г., 01:50 Liz Baron <notifications@github.com>:

> Hi @dkultasev <https://github.com/dkultasev>,
>
> Sebastian is temporarily unavailable, but did you try what he suggested?
> The tSQLt build takes care of building the Visual Studio solution. There is
> no additional step required. And it looks like the error is pretty clear,
> and is specific to your local setup.
>
> If that is not where you are stuck anymore, can you please give us details
> on:
>
>    - What are you trying to do?
>    - What problem did you encounter?
>    - How have you tried to fix it?
>
> Let me know!
> Liz
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tSQLt-org/tSQLt/pull/53#issuecomment-719087631>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGCNCYEAWLPYQGCWFGK5R33SNH5TXANCNFSM4JWNZJKA>
> .
>

dkultasev commented:
hi @lizbaron, the issue was the trailing slash symbol at the end of the Msbuild folder path. Now I'm getting the certificate problem, will try to fix that later:

```
C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Csc.exe /noconfig /nowarn:1701,1702 /nostdlib+ /platform:AnyCPU /errorreport:prompt /define:TRACE /highentropyva- /reference:C:\Windows\Microsoft.NET\Framework64\v2.0.50727\mscorlib.dll /reference:C:\Windows\Microsoft.NET\Framework64\v2.0.50727\System.Data.dll /reference:C:\Windows\Microsoft.NET\Framework64\v2.0.50727\System.dll /reference:C:\Windows\Microsoft.NET\Framework64\v2.0.50727\System.Transactions.dll /reference:C:\Windows\Microsoft.NET\Framework64\v2.0.50727\System.Xml.dll /debug:pdbonly /optimize+ /out:obj\CruiseControl\tSQLtCLR.dll /target:library /utf8output Annotations.cs AssemblyCopyright.cs AssemblyInfo.cs CommandExecutor.cs CommandExecutorException.cs OutputCaptor.cs tSQLtPrivate.cs InvalidResultSetException.cs MetaDataEqualityAsserter.cs ResultSetFilter.cs StoredProcedures.cs TestDatabaseFacade.cs
     [exec] AssemblyInfo.cs(35,12): warning CS1699: Use command line option '/keycontainer' or appropriate project settings instead of 'System.Reflection.AssemblyKeyNameAttribute' [C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\tSQLtCLR.csproj]
     [exec] CSC : error CS1548: Cryptographic failure while signing assembly 'c:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\obj\CruiseControl\tSQLtCLR.dll' -- 'The key container name 'tSQLt_OfficialSigningKey' does not exist' [C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\tSQLtCLR.csproj]
     [exec] Done Building Project "C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\tSQLtCLR.csproj" (default targets) -- FAILED.
     [exec] 
     [exec] Build FAILED.
     [exec] 
     [exec] "C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\tSQLtCLR.csproj" (default target) (1) ->
     [exec] (CoreCompile target) -> 
     [exec]   AssemblyInfo.cs(35,12): warning CS1699: Use command line option '/keycontainer' or appropriate project settings instead of 'System.Reflection.AssemblyKeyNameAttribute' [C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\tSQLtCLR.csproj]
     [exec] 
     [exec] 
     [exec] "C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\tSQLtCLR.csproj" (default target) (1) ->
     [exec] (CoreCompile target) -> 
     [exec]   CSC : error CS1548: Cryptographic failure while signing assembly 'c:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\obj\CruiseControl\tSQLtCLR.dll' -- 'The key container name 'tSQLt_OfficialSigningKey' does not exist' [C:\Users\dkult\OneDrive\Docs\Sources\tSQLt-1\tSQLtCLR\tSQLtCLR\tSQLtCLR.csproj]
     [exec] 
     [exec]     1 Warning(s)
     [exec]     1 Error(s)
```

dkultasev commented:
I'm still struggling on trying to run the build locally. The error is still the same. This is what I've tried so far:
* Created certificate (no password)
```
"C:\Program Files\OpenSSL-Win64\bin\openssl.exe" req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout key.pem -out cert.pem
"C:\Program Files\OpenSSL-Win64\bin\openssl.exe" pkcs12 -inkey key.pem -in cert.pem -export -out tSQLtOfficialSigningKey.pfx
```
* Copied `tSQLtOfficialSigningKey.pfx` file to `tSQLtCLR\OfficialSigningKey`
* Executed `Import-PfxCert.ps1` (seems that it passed)
* Executed `InstallSigningKey.bat` file (it asked for password)

What should I do?

dkultasev commented:
If I execute `certutil -csp "Microsoft Strong Cryptographic Provider" -key` I see the entry there:
```
  tSQLt_OfficialSigningKey
  d140ecbbd4084f5c98864a945124ad5b_7ab133a0-7368-497d-b89f-aaaa66dcbce5
  RSA
    AT_KEYEXCHANGE
```

dkultasev commented:
@lizbaron, @mbt1  please check. All tests passed. I've added new test for CLR function.

mbt1 commented on Source/tSQLt.Private_PrepareFakeFunctionOutputTable.ssp.sql:
```diff
@@ -0,0 +1,26 @@
+             AND OBJECT_ID(@FakeDataSource) IS NULL
+           )
+            BEGIN
+                SET @FakeDataSource = CONCAT(N'(', @FakeDataSource, N') a');
```
We are still supporting 2008R2. CONCAT is not available in that version.
([https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms181984(v=sql.105)](https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms181984(v=sql.105)))

mbt1 commented on Source/tSQLt.Private_PrepareFakeFunctionOutputTable.ssp.sql:
```diff
@@ -0,0 +1,26 @@
+                SET @FakeDataSource = CONCAT(N'(', @FakeDataSource, N') a');
+            END;
+
+        DECLARE @Cmd NVARCHAR(MAX) = CONCAT('SELECT * INTO ', @OutputTable, ' FROM ', @FakeDataSource);
```
see above

mbt1 commented on Tests/FakeFunctionTests.class.sql:
```diff
@@ -422,6 +431,131 @@ BEGIN
   EXEC FakeFunctionTests.[assert parameter missmatch causes error];
 END;
 GO
+CREATE PROCEDURE FakeFunctionTests.[test errors when both @FakeFunctionName and @FakeDataSource are passed]
```
It looks like this test name and the next test's name are flipped.

mbt1 commented on Tests/FakeFunctionTests.class.sql:
```diff
@@ -422,6 +431,131 @@ BEGIN
+  EXEC tSQLt.FakeFunction @FunctionName = 'FakeFunctionTests.AFunction', @FakeFunctionName = 'FakeFunctionTests.AFunction', @FakeDataSource = 'select 1 one';
+END;
+GO
+CREATE PROCEDURE FakeFunctionTests.[test errors when scalar function is passed as @FakeDataSource]
```
The test name is unclear. How about [test errors if function is a SVF and @FakeDataSource is used]?

In addition: We need a similar test for CLR SVF. I did not see that.

mbt1 commented on Tests/FakeFunctionTests.class.sql:
```diff
@@ -181,6 +181,15 @@ BEGIN
   EXEC FakeFunctionTests.[Assert errors on function type mismatch] 'FakeFunctionTests.[An SVF]', 'FakeFunctionTests.[A Table]';
 END;
 GO
+CREATE PROCEDURE FakeFunctionTests.[test errors when function is SVF, fake is CLRTVF and data source is used]
```
I'm not sure what this test is trying to test. The general behavior should be:

1. Any TVF can be faked using @FakeDataSource [3 test scenarios: MS-TVF, IL-TVF, CLR-TVF]
1. Any SVF cannot (yet) be faked using @FakeDataSource [2 test scenarios: SVF, CLR-SVF] (I would like to talk about this part later, but for now, let's get this PR approved first.)

This test's name does not reflect what the test is doing, and the test seems to not align with those requirements listed above.

dkultasev replied:
I made requested fixes. I've added the test for CLR-TVF. Following test was removed as probably I misunderstood what was supposed to b e done.
Is there anything else I need to update/add?

mbt1 reviewed (approved):
Great job, Dmitrij. I'm approving this PR.
